File: 221-maximal-square/maximal-square.cpp
class Solution {
public:
    int maximalSquare(vector<vector<char>>& matrix) {
        int m = matrix.size();
        int n = matrix[0].size();
        vector<vector<int>> dp(m + 1, vector<int>(n + 1, 0));
        int max_length = 0;
        for(int i=m-1; i>-1; i--){
            for(int j=n-1; j>-1; j--){
                if(matrix[i][j]=='1'){
                    dp[i][j] = 1 + min({dp[i+1][j], dp[i][j+1], dp[i+1][j+1]});
                    max_length=max(max_length,dp[i][j]);
                }
            }
        }
        return max_length*max_length;
    }
};

File: 221-maximal-square/maximal-square.py
class Solution:
    def maximalSquare(self, matrix: List[List[str]]) -> int:
        rows=len(matrix)
        cols=len(matrix[0])
        cache={}
        def helper(r,c):
            if r>=rows or r<0 or c>=cols or c<0:
                return 0
            if (r,c) not in cache:
                down=helper(r+1,c)
                right=helper(r,c+1)
                diag=helper(r+1,c+1)
                cache[(r,c)]=0
                if matrix[r][c]=="1":
                    cache[(r,c)]= 1+ min(down,diag,right)
            return cache[(r,c)]

        helper(0,0)
        return max(cache.values())**2